### DISCUSSION TITLE
How will we manage dependencies across plugins?

### DISCUSSION THREAD

**Discussion by jondavid-black** [DevOps-MBSE/AaC#67] [Q&A]
We need a reliable way to address dependencies among plugins.

Could we use something like a requirements.txt approach?

**Comment by Coffee2Bits** [comment]
### Considerations

1. Plugins depending on the output of other plugins (Think doc generation requiring plantUML diagrams to be generated)
2. Since plugins can define extensions, there will be AaC models that are invalid unless they are parsed with the extensions from the plugins they rely on

### Solutions

#### Plugin Dependecy requirement
Plugins can currently define their plugins dependencies using setup.py and pip install. So, for plugins with dependencies on other plugins, leveraging the setup.py should be sufficient once we've started publishing our plugins to PyPi.

#### AaC File Plugin Requirements 
As for yaml files that rely on plugins, the only solution I can think of is either letting the users determine all present plugins with a setup.py in their AaC projects, or we add a root node that defines plugins/versions required by the AaC file so those requirements live with the AaC file.

**Reply by Coffee2Bits** [reply]
I'm going to leave a quick comment as a note/reminder of this idea. 

Since the AaC DSL is self-defining and the plugins leverage the AaC DSL, we could effectively 'remove' plugin extension dependencies by injecting the plugin's data/enums definitions into the model's file. So, if `example_arch.yaml` uses extensions from say the protobuf plugin, we could make the `example_arch.yaml` a valid architecutre/definition without the protobuf plugin by injecting the protobuf plugin's AaC definitions/extensions into the `example_arch.yaml` file.

Obviously this solution isn't very helpful since the plugin will probably just provide plugin-specific definitions and extensions, but I wanted to document this while I think about it some more.

**Comment by jondavid-black** [comment]
Another consideration to discuss is if/how we may want to establish a "core plugin set" and an ability to "promote" a plugin into that set.  I'm thinking about language standards in other realms like Java and how they use JSRs to enhance the language over time.  Perhaps we could use plugins in a similar manner.  Another analogy could be Minecraft plugins.  My kids love to play using plugins that add new aspects to the game.  There are launchers that allow then to "check the right boxes" to get the configuration they want.  Maybe a similar "AaC Launcher" or "AaC Installer" would enable a desired outcome.

That raises a question of whether the core AaC application should have more inherent functionality.  Thoughts?

We certainly need to address dependencies, but we cannot overlook the potential for conflicts.  Should we provide a means of designating known plugin incompatibility?  If encountered, what is the appropriate user action...if any?

**Reply by Coffee2Bits** [reply]
I like the idea of a core plugin set -- I personally prefer to use the optional tools/extensions/plugins that are maintained by core/official teams, especially if the ecosystem isn't mature or very extensive. I think we should explore what a core plugin set would look like to us.

That's a fantastic point about mod-loaders/managers. Having used them for modding various games, this is definitely a must have in my opinion. Users do have a fairly easy way to install/uninstall plugins via pip, but we could add a ton of value if we could provide things like plugin profiles (a list of active/inactive plugins to use) or even import/export active plugin lists for users to share/manage with with AaC files. 

The pattern that's emerging is that the AaC project is providing core functionality that's leveraged by plugin developers to create tangible user features. I think maintaining a core plugin set that is compatible/well-kept will be sufficient unless I'm misunderstanding your question, "That raises a question of whether the core AaC application should have more inherent functionality."

Unfortunately, I don't think that there is a good way to manage incompatibility between plugins, and I'm failing to think of what a conflict might look like except for name clashes between plugin extensions/definitions. We could add an AaC field for known incompatibility between plugins and run a check during start-up for those known incompatibilities, but I think it would have to ultimately be on the user to determine if the conflict is an issue and whether or not to deactivate one of the plugins in conflict.

**Comment by lizzcondrey** [comment]
I think some of this as come back around with the recent discussion thread #670? Or am I improperly mixing these? There is definitely additional captured here that has not come up in the #670 thread, but the initial issue of addressing cross plugin dependencies is in both, I think.

**Comment by jondavid-black** [comment]
I think is is absolutely coming back...and suggest the following.

Just allow dependent plugins to be declared within the plugin definition for a given command.  It may look something like:

```yaml
plugin:
  name: My Plugin
  description: An AaC plugin that folds your laundry and does your dishes.
  commands:
    - name: fold-laundry
      helpText: Tackles your laundry.
      run_before:
        - check
      run_after:
        - check
```

**Reply by lizzcondrey** [reply]
Any issue with me combining the two and shuffling the information one way or the other to keep things together and more organized then?
